### PR TITLE
Fixed Media Layout in RTL, fix #2101.

### LINKED ISSUE #2101: Media layout doesn't work as expected in RTL locales
First identified by @vrjgamer and partially addressed by #1958.

The layout should be mirrored from LTR.

### PR DESCRIPTION
**Pull request by huangyz0918**
Closes #2101 

#### What has been done to verify that this works as intended?
Despite of API 16, I made some changes in `MediaLayout.java` and make it mirrored from LTR.

Screen Records in different APIs:

| Device API | LTR  | RTL |
| :------------: |:---------------:|:-----:|
| Android 4.4      |  |  |
| Android 5.0      |  |  |
| Android 6.0      |  |  |
| Android 7.1      |  |  |
| Android 8.1      |  |  |

#### Why is this the best possible solution? Were any other approaches considered?
It's the best solution without a refactor but not the most readable code, if any one have spare time, welcome to refactor that ! 👍 ——> issue #2127 
#### Are there any risks to merging this code? If so, what are they?
No.
#### Do we need any specific form for testing your changes? If so, please attach one.
Recommend `Birds.xml` or something else.
#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
No.

### CODE REVIEW

**Comment by huangyz0918** [comment]
Known issues：

1. Cannot cover **API level 16** ( I just skipped  that ).

2. Android 4.2 ( API level 17 ) **has some unnecessary margins** in media buttons.

**Comment by huangyz0918** [comment]
Like @grzesiek2010 said, we should build an adapter for those all lists : )
So I changed the PR name to do refactor and fix the media layout in a single one.
Thanks for every one's suggestion, it's a great way to learn 👏

**Comment by huangyz0918** [comment]
@grzesiek2010 @lognaturel Please review, thank you.

**Comment by huangyz0918** [comment]
@opendatakit-bot  label "needs review"

**Comment by huangyz0918** [comment]
@grzesiek2010 @lognaturel Please review if you have spare time.
Thank you ☺️

File: collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java [hunk 1]
```diff
@@ -362,35 +359,30 @@ public void onClick(View v) {
             //
             // Text upper left; audio upper right; video below audio on right.
             // image below text, audio and video buttons; left-aligned with text.
-            textParams.addRule(RelativeLayout.ALIGN_PARENT_LEFT);
+            if (isNeedPatchRTL()) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: ̶I̶ ̶d̶o̶n̶'̶t̶ ̶u̶n̶d̶e̶r̶s̶t̶a̶n̶d̶ ̶t̶h̶i̶s̶ ̶f̶r̶a̶g̶m̶e̶n̶t̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶S̶T̶A̶R̶T̶`̶ ̶a̶n̶d̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶L̶E̶F̶T̶`̶ ̶s̶h̶o̶u̶l̶d̶ ̶b̶e̶ ̶t̶h̶e̶ ̶s̶a̶m̶e̶ ̶s̶o̶ ̶t̶h̶e̶y̶ ̶a̶r̶e̶ ̶u̶s̶e̶d̶ ̶u̶n̶d̶e̶r̶ ̶d̶i̶f̶f̶e̶r̶e̶n̶t̶ ̶c̶i̶r̶c̶u̶m̶s̶t̶a̶n̶c̶e̶s̶?̶

File: collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java [hunk 2]
```diff
@@ -411,6 +403,40 @@ public void onClick(View v) {
...
+    }
+
+    /**
+     * used for inject layout rules by locales
+     */
+    private void injectRulesByLocale(RelativeLayout.LayoutParams mediaParams,
+                                     RelativeLayout.LayoutParams imageParams,
+                                     int imagePadding) {
+        mediaParams.addRule(RelativeLayout.ALIGN_PARENT_TOP);
+        if (isNeedPatchRTL()) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: ̶T̶h̶e̶ ̶s̶a̶m̶e̶ ̶a̶s̶ ̶a̶b̶o̶v̶e̶ ̶b̶u̶t̶ ̶h̶e̶r̶e̶ ̶w̶i̶t̶h̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶E̶N̶D̶`̶ ̶a̶n̶d̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶R̶I̶G̶H̶T̶`̶

File: collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java [hunk 3]
```diff
@@ -411,6 +403,40 @@ public void onClick(View v) {
...
+        } else {
+            mediaParams.setMargins(5, 0, 11, 0);
+            mediaParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT);
+            imageParams.addRule(RelativeLayout.LEFT_OF, imagePadding);
+        }
+    }
+
+    // used for items without image
+    private void injectRulesByLocale(RelativeLayout.LayoutParams mediaParams) {
+        if (isNeedPatchRTL()) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: ̶A̶n̶d̶ ̶o̶n̶e̶ ̶m̶o̶r̶e̶ ̶t̶i̶m̶e̶.̶

File: collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java [hunk 4]
```diff
@@ -302,6 +304,9 @@ public void onClick(View v) {
 
         // Determine the layout constraints...
         // Assumes LTR, TTB reading bias!
+        int startOf = isNeedPatchRTL() ? RelativeLayout.ALIGN_PARENT_START : RelativeLayout.ALIGN_PARENT_LEFT;
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: ̶a̶n̶d̶ ̶t̶h̶e̶ ̶s̶a̶m̶e̶ ̶h̶e̶r̶e̶.̶

**Comment by grzesiek2010** [comment]
Ok, you can ignore my comments.
Left is always left but strat is left or right depending on the language of course.

File: collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java (hunk 1 above)
Comment (huangyz0918) [reply collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: I'm sorry, do you mean using `ALIGN_PARENT_START` and `ALIGN_PARENT_LEFT` will lead to the same result? (in both RTL and LTR)

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: No, I just had a temporary black hole in my head and didn't catch it.

File: collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java [hunk 5]
```diff
@@ -302,6 +304,8 @@ public void onClick(View v) {
 
         // Determine the layout constraints...
         // Assumes LTR, TTB reading bias!
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java]: This comment is no longer accurate! How about changing it to "Determine the layout constraints. Right-to-left support works for API > 16."

**Review by lognaturel — APPROVED** [review]

**Comment by huangyz0918** [comment]
@opendatakit-bot unlabel "needs review"

**Comment by mmarciniak90** [comment]
@huangyz0918, @lognaturel I do not feel like a specialist on RTL languages.
I agree that view with media is mirrored from LTR.

|master | fix LTR| fix RTL| 
|---|---|---|
|screenshot_20180521-105412|screenshot_20180521-115114|screenshot_20180521-121045|

I'm not sure that this fix should cause changes on this view too:

|master |fix |
|---|---|
|screenshot_20180521-105240|screenshot_20180521-115054|

* Group brackets are incorect on Android 4.4 and 4.2
* First question is not fully visible. Speaker icon covers question mark.
screenshot_2018-05-21-12-37-02
* No changes on Android 4.1
screenshot_2018-05-21-11-47-59

**Comment by huangyz0918** [comment]
> I'm not sure that this fix should cause changes on this view too:

I can fix this.

> Group brackets are incorect on Android 4.4 and 4.2

And this (but I'm not sure this is a common issue since I cannot reproduce it in Android 4.4, you can see in my screenshot).

> No changes on Android 4.1

We just gave up API 16 since it's not supported by RTL.

**Comment by huangyz0918** [comment]
@mmarciniak90 I made some improvements.
> I'm not sure that this fix should cause changes on this view too:

I made this view mirrored from LTR too. (> android 4.2)

> Group brackets are incorect on Android 4.4 and 4.2

Fixed. 

> First question is not fully visible. Speaker icon covers question mark.

Sorry, but I cannot reproduce that (I adjusted the margins and hope it works).

**Comment by mmarciniak90** [comment]
**Tested with success**

Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1
Mentioned problems were fixed. @huangyz0918 great job! :1st_place_medal: 
Verified on light and dark mode. 

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Pull request merged**